File: Dev/unitTest_Engine_cpp_gtest/BugFix/DisposeAndChangeParent.cpp
#include <gtest/gtest.h>
#include <Altseed.h>
#include "../EngineTest.h"

using namespace std;
using namespace asd;

class BugFix_DisposeAndChangeParent : public EngineTest
{
public:
	BugFix_DisposeAndChangeParent(bool isOpenGLMode)
		: EngineTest(asd::ToAString("DisposeAndChangeParent"), isOpenGLMode, 120)
	{
	}

protected:
	TextureObject2D::Ptr child, parent1, parent2, parent3;

	void OnStart()
	{
		auto texture = GetTexture();

		child = make_shared<TextureObject2D>();
		child->SetTexture(texture);
		child->SetPosition(Vector2DF(300, 0));
		child->SetScale(Vector2DF(0.4f, 0.4f));

		parent1 = make_shared<TextureObject2D>();
		parent1->SetTexture(texture);
		parent1->SetPosition(Vector2DF(0, 0));
		parent1->SetScale(Vector2DF(0.4f, 0.4f));
		parent1->SetColor(Color(255, 0, 0, 255));

		parent2 = make_shared<TextureObject2D>();
		parent2->SetTexture(texture);
		parent2->SetPosition(Vector2DF(0, 128));
		parent2->SetScale(Vector2DF(0.4f, 0.4f));
		parent2->SetColor(Color(0, 255, 0, 255));

		parent3 = make_shared<TextureObject2D>();
		parent3->SetTexture(texture);
		parent3->SetPosition(Vector2DF(0, 256));
		parent3->SetScale(Vector2DF(0.4f, 0.4f));
		parent3->SetColor(Color(0, 0, 255, 255));

		parent1->AddChild(child, ChildManagementMode::Nothing, ChildTransformingMode::Position);

		Engine::AddObject2D(child);
		Engine::AddObject2D(parent1);
		Engine::AddObject2D(parent2);
		Engine::AddObject2D(parent3);
	}

	void OnUpdating()
	{
		if (m_currentTime == 30)
		{
			parent1->Dispose();
			parent2->AddChild(child, ChildManagementMode::Nothing, ChildTransformingMode::Position);
		}
		if (m_currentTime == 60)
		{
			if (child->GetParent() != nullptr)
			{
				child->GetParent()->RemoveChild(child);
			}
			parent3->AddChild(child, ChildManagementMode::Nothing, ChildTransformingMode::Position);
		}
		if (m_currentTime == 90)
		{
			if (child->GetParent() != nullptr)
			{
				child->GetParent()->RemoveChild(child);
			}
			parent2->AddChild(child, ChildManagementMode::Nothing, ChildTransformingMode::Position);
		}
	}

	void OnUpdated()
	{
	}

	void OnFinish()
	{
	}
};

ENGINE_TEST(BugFix, DisposeAndChangeParent)

File: Dev/unitTest_Engine_cs/BugFix/DisposeAndChangeParent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using asd;

namespace unitTest_Engine_cs.BugFix
{
	class DisposeAndChangeParent : EngineTest
	{
		TextureObject2D child, parent1, parent2, parent3;

		public DisposeAndChangeParent()
			: base(120)
		{
		}

		protected override void OnStart()
		{
			Texture2D texture = Engine.Graphics.CreateTexture2D(CloudTexturePath);
			parent1 = new TextureObject2D()
			{
				Texture = texture,
				Position = new Vector2DF(0, 40),
				Color = new Color(0, 255, 0),
			};
			parent2 = new TextureObject2D()
			{
				Texture = texture,
				Position = new Vector2DF(0, 80),
				Color = new Color(128, 128, 128),
			};
			parent3 = new TextureObject2D()
			{
				Texture = texture,
				Position = new Vector2DF(0, 120),
				Color = new Color(128, 0, 0),
			};
			child = new TextureObject2D()
			{
				Texture = texture,
				Position = new Vector2DF(450, 0),
			};

			parent1.AddChild(child, ChildManagementMode.Nothing, ChildTransformingMode.Position);

			Engine.AddObject2D(parent1);
			Engine.AddObject2D(parent2);
			Engine.AddObject2D(parent3);
			Engine.AddObject2D(child);
		}

		protected override void OnUpdating()
		{
			if(Time == 30)
			{
				parent1.Dispose();
				parent2.AddChild(child, ChildManagementMode.Nothing, ChildTransformingMode.Position);
			}
			if(Time == 60)
			{
				if(child.Parent != null)
				{
					child.Parent.RemoveChild(child);
				}
				parent3.AddChild(child, ChildManagementMode.Nothing, ChildTransformingMode.Position);
			}
			if(Time == 90)
			{
				if(child.Parent != null)
				{
					child.Parent.RemoveChild(child); 
				}
				parent2.AddChild(child, ChildManagementMode.Nothing, ChildTransformingMode.Position);
			}
		}
	}
}
